Public/Src/Tools/Execution.Analyzer/Analyzers.Core/XLGPlusPlus/XldbProtobufExtensions.cs

```diff
@@ -558,5 +548,479 @@ public static Xldb.Fingerprint ToFingerprint(this Fingerprint fingerprint)
+        {
+            var xldbPipTable = new Xldb.PipTable
+            {
+                IsDisposed = pipTable.IsDisposed,
+                Reads = pipTable.Reads,
+                Writes = pipTable.Writes,
+                Count = pipTable.Count,
+                PageStreamsCount = pipTable.PageStreamsCount,
+                Size = pipTable.Size,
+                WritesMilliseconds = pipTable.WritesMilliseconds,
+                ReadsMilliseconds = pipTable.ReadsMilliseconds,
+                Used = pipTable.Used,
+                Alive = pipTable.Alive
+            };
+
+            xldbPipTable.StableKeys.AddRange(pipTable.StableKeys.Select(stableKey => stableKey.Value));
+            xldbPipTable.Keys.AddRange(pipTable.Keys.Select(key => key.Value));
+            xldbPipTable.DeserializationContexts.AddRange(pipTable.DeserializationContexts.Select(
+                context => new PipDeserializationContext() { Key = (PipQueryContext)context.Key, Value = context.Value }));
+            return xldbPipTable;
+        }
+
+        /// <nodoc />
+        public static Xldb.PipData ToPipData(this PipData pipData)
+        {
+            return !pipData.IsValid ? null : new Xldb.PipData
+            {
+                IsValid = pipData.IsValid,
+                FragmentSeparator = pipData.FragmentSeparator.ToString(),
+                FragmentCount = pipData.FragmentCount,
+                FragmentEscaping = (PipDataFragmentEscaping)pipData.FragmentEscaping
+            };
+
+        }
+
+        /// <nodoc />
+        public static Xldb.PipProvenance ToPipProvenance(this PipProvenance provenance)
+        {
+            return provenance == null ? null : new Xldb.PipProvenance()
+            {
+                Usage = provenance.Usage.ToPipData(),
+                ModuleId = provenance.ModuleId.Value.Value,
+                ModuleName = provenance.ModuleName.ToString(),
+                SemiStableHash = provenance.SemiStableHash
+            };
+        }
+
+        public static Xldb.FileOrDirectoryArtifact ToFileOrDirectoryArtifact(this FileOrDirectoryArtifact artifact, PathTable pathTable)
+        {
+            var xldbFileOrDirectoryArtifact = new Xldb.FileOrDirectoryArtifact()
+            {
+                IsValid = artifact.IsValid
+            };
+
+            if (artifact.IsDirectory)
+            {
+                xldbFileOrDirectoryArtifact.IsDirectory = true;
+                xldbFileOrDirectoryArtifact.DirectoryArtifact = artifact.DirectoryArtifact.ToDirectoryArtifact(pathTable);
+            }
+            else
+            {
+                xldbFileOrDirectoryArtifact.IsFile = true;
+                xldbFileOrDirectoryArtifact.FileArtifact = artifact.FileArtifact.ToFileArtifact(pathTable);
+            }
+
+            return xldbFileOrDirectoryArtifact;
+        }
+
+        /// <nodoc />
+        public static Xldb.Pip ToPip(this Pip pip, DirectedGraph directedGraph)

```

Comment:
>Xldb.Pip [](start = 22, length = 8)

I see there are specific pip objects for all relevant pip types (Process, CopyFile, ...) How do you see Xldb.Pip being used?

Reply:
It is used as the ParentPipInfo object for each of the specific pips (since Protobuf doesn't support inheritance, but we still want this information stored in the specific pips)

---
In reply to: [314446991](https://github.com/microsoft/BuildXL/pull/724#discussion_r314446991) [](ancestors = 314446991)

Reply:
Then we should make sure none of the fields like Provenance are overlapping

It doesn't promote consistency but I wonder if it would be a nicer experience to just push the useful fields from Pip onto Process and the othe rspecific pip types. I think some of those fields won't be to all pip types

---
In reply to: [314450315](https://github.com/microsoft/BuildXL/pull/724#discussion_r314450315) [](ancestors = 314450315,314446991)

Reply:
I removed overlapping (overridden) fields so that the specific pips contain the information and not the "parentPip" object. However I think this abstraction of having info like the semistable hash/pipid and the edges stored in this ParentObject make it easier to understand that these are fields found in pips of any type

---
In reply to: [314453819](https://github.com/microsoft/BuildXL/pull/724#discussion_r314453819) [](ancestors = 314453819,314450315,314446991)